### ISSUE TITLE
Add a 'notes' field to persons

### ISSUE THREAD

**Issue by gvwilson** [swcarpentry/amy#323]
We may want to record extra information about instructors - @tracykteal has requested.

**Comment by wking** [comment]
On Thu, May 28, 2015 at 08:17:57AM -0700, Greg Wilson wrote:
> We may want to record extra information about instructors -
> @tracykteal has requested.

This sounds reasonable to me, but I find it useful to at least have a
guess at the sort of information you want to store before adding a
field like this.  For example, see [1], where I argue for a structured
approach to storing waitlist-y information.

[1]: http://lists.software-carpentry.org/pipermail/amy_lists.software-carpentry.org/2015-February/000064.html

**Comment by gvwilson** [comment]
The problem is that we don't know what we'll want to store - is someone 
only able to teach outside the academic year in Sweden?  Do they have a 
nut allergy?  Etc.

**Comment by tracykteal** [comment]
Right, the idea for this field is that it's the miscellaneous information that we want to keep track of, so basically a free text field. Is this reasonable?

**Comment by wking** [comment]
On Thu, May 28, 2015 at 12:06:55PM -0700, Greg Wilson wrote:
> The problem is that we don't know what we'll want to store - is
> someone only able to teach outside the academic year in Sweden?  Do
> they have a nut allergy?  Etc.

So before adding any data like that to an unstructured field, I want
to understand what the workflow around it is going to look like.  If
we store nut allergies in Person.notes, how does that get used?  Is it
someone's job to go through and check the notes fields before each
event to look for possible allergy conditions?  Maybe we want an
'Accessability' table that's many-to-many for people and sites, so we
can automatically check that we aren't assigning elevator-requiring
instructors to stair-only sites.  That's a bit trickier for one-off
things like menus (I don't expect nut-factory sites), but it would at
least make it easy for a host to see the union of all instructor (and
student?) accessability requirements automatically.

For “can only teach outside the academic year in Sweden”, I'd consider
modeling free/busy time following RFC 3.6.4 [1] and then feed that
into an iCalendar library to figure out which instructors were free
for a particular time slot.

[1]: https://tools.ietf.org/html/rfc5545#section-3.6.4

**Comment by wking** [comment]
On Thu, May 28, 2015 at 12:19:46PM -0700, Tracy Teal wrote:
> Right, the idea for this field is that it's the miscellaneous
> information that we want to keep track of, so basically a free text
> field. Is this reasonable?

It's easy to add, but in almost every scenario I play through in my
mind an unstructured field turns into a dumping ground for information
that's easy to forget and hard to work with ;).  It's basically giving
up on ever automating a particular class of data, and then requiring
admins to manually review and synthesize that unstructured data by
remembering to follow whatever the note-adder's intended workflow was.
So if we don't want to structure the data for Amy to look at, we at
least we need a policy to say “admins should review Person.notes for a
given person when …” so we know when the admins should be looking at
that field.

**Comment by tracykteal** [comment]
My current use is case is that someone is interested in teaching in Taiwan when she goes to visit home, so I want to somehow note that. Or someone might be interested in helping to develop lessons in social science or translating lessons to Mandarin. I envision it as sort of a 'grep' field.

**Comment by wking** [comment]
On Thu, May 28, 2015 at 12:27:36PM -0700, Tracy Teal wrote:
> My current use is case is that someone is interested in teaching in
> Taiwan when she goes to visit home, so I want to somehow note that.

So following the structured free/busy approach, we'd want a table with
a key into Person, a key into Airport, and a free/busy entry.  Then
Amy could give you a list of all instructors who were free at an
airport within $x miles of your site on a given date (and allow you to
also filter by skill, previous teaching count, etc.).  Searching for
all instructors with unstructured infromation about Taiwan would
probably work, but I'm less optimistic about searching for instructors
with unstructured information about New York or another of our
high-density locations.

> Or someone might be interested in helping to develop lessons in
> social science or translating lessons to Mandarin. I envision it as
> sort of a 'grep' field.

I don't think we want to be storing information about
lesson-development here [1], since this is about organizing events.
I'd store that information in a lesson-managment-specific location
(currently maybe just [2])

[1]: https://github.com/swcarpentry/amy/issues/301#issuecomment-105107198
[2]: https://github.com/swcarpentry/site/blob/master/lessons.html

**Comment by wking** [comment]
On Thu, May 28, 2015 at 12:25:17PM -0700, W. Trevor King wrote:
> It's easy to add, but in almost every scenario I play through in my
> mind an unstructured field turns into a dumping ground for
> information that's easy to forget and hard to work with ;).

Just to re-enforce this point, I'm pushing for structured data to
shift the burden of managing the data off the admins and onto the devs
and Amy's CPUs ;).  If my proposals don't sound like they'd save admin
time, it would certainly be less dev time to just add a notes field
and let the admins manage/consume it by hand.

**Comment by gvwilson** [comment]
I vote for a free-form text field - @tracykteal?
Cheers,
G

**Comment by pbanaszkiewicz** [comment]
@wking:
> So following the structured free/busy approach, we'd want a table with a key into Person, a key into Airport, and a free/busy entry.

Technologically it's feasible, but I don't think we could gather accurate occupation data and keep it up-to-date.

**Comment by wking** [comment]
On Thu, May 28, 2015 at 01:38:22PM -0700, Piotr Banaszkiewicz wrote:
> > So following the structured free/busy approach, we'd want a table
> > with a key into Person, a key into Airport, and a free/busy entry.
> 
> Technologically it's feasible, but I don't think we could gather
> accurate occupation data and keep it up-to-date.

Keeping the data up to date seems orthogonal to
structured/unstructured to me.  Is the idea that admins parsing the
unstructured data by eye will intuitively know when it's stale?

**Comment by pbanaszkiewicz** [comment]
@wking:

> Keeping the data up to date seems orthogonal to structured/unstructured to me.

It doesn't make sense to create a feature that is impossible to rely on. Free/busy statuses change very quickly (for example I can't plan ahead my schedule for more than a week). So it's:

1. hard: to hook into people's calendars
2. harder: to get people to update their calendars regularly.

> Is the idea that admins parsing the unstructured data by eye will intuitively know when it's stale?

I believe free/busy statuses aren't stored anywhere. I also believe this is the kind of data that gets obsolete most often.

I think admins can tell stale from fresh data by eye-parsing timestamped data (either unstructured or structured). For example: `"2014.10 Piotr expressed a desire to teach Git in January 2015"` (probably stale), `"2015.04.27 Piotr got into GSOC'15, he should add awards, filtering, autocompletion, and reports to Amy"` (probably fresh).

To sum up: I don't think we should gather free/busy data from our instructors.

**Comment by wking** [comment]
On Fri, May 29, 2015 at 12:48:01AM -0700, Piotr Banaszkiewicz wrote:
> I believe free/busy statuses aren't stored anywhere. I also believe
> this is the kind of data that gets obsolete most often.

I agree that it's going to be hard to keep this up to date if Amy
isn't pulling the information directly from a person's external
calendar.

> I think admins can tell stale from fresh data by eye-parsing
> timestamped data (either unstructured or structured). For example:
> `"2014.10 Piotr expressed a desire to teach Git in January 2015"`
> (probably stale), `"2015.04.27 Piotr got into GSOC'15, he should add
> awards, filtering, autocompletion, and reports to Amy"` (probably
> fresh).
> 
> To sum up: I don't think we should gather free/busy data from our
> instructors.

Well, two of the handful of examples we have motivating Person.notes
are about free/busy data (“can only teach outside the academic year in
Sweden” [1] and “interested in teaching in Taiwan when she goes to
visit home” [2]).  If we had timestamped notes (or a searchable
HyperKitty archive (#299)?  I haven't checked for an API for this),
would consuming that information look like:

  Show me all instructors who's nearest airport is within 50 miles of
  TPE or who have archived emails with “Taiwan” in the body and dates
  in the last six months.

And then looking through (or inlining?) those emails to let the admin
decide whether the instructors are worth contacting about a workshop
in Taipei?  I'd rather just have the parsing to free/busy times happen
at email-receipt time (if the instructors don't publish free/busy
information using iCalendar) and then use Django for the subsequent
queries.  Or just not bother with this information at all.  But
offloading the parsing to admin-searching-for-instructors-time doesn't
seem like the most efficient solution, if only because it requires
admins to read emails/notes at least twice (once when they arrive, and
then again during their instructor-search).

[1]: https://github.com/swcarpentry/amy/issues/323#issuecomment-106568256
[2]: https://github.com/swcarpentry/amy/issues/323#issuecomment-106568256

**Comment by tracykteal** [comment]
I think we're expanding the use case for the Persons information to include information not only that an admin would want for scheduling workshops, but also information that people like EDs would want. We want to know things like "would like to write a grant on astronomy with us in the next NSF round". We certainly can keep this information somewhere else, but it seems like it would be useful to integrate it into Amy. 

I also agree that Free/Busy will be basically impossible to keep up to date.

**Comment by pbanaszkiewicz** [comment]
Closed via #327.
